app/scripts/controllers/network/network-controller.ts

```diff
@@ -154,20 +154,22 @@ type NetworkControllerEvent =
 /**
  * The messenger that the NetworkController uses to publish events.
  */
-type NetworkControllerMessenger = RestrictedControllerMessenger<
+export type NetworkControllerMessenger = RestrictedControllerMessenger<
   typeof name,
   never,
   NetworkControllerEvent,
   never,
-  NetworkControllerEventType
+  // TODO: Why is this `string` and not `NetworkControllerEventType`?
+  // Also this breaks `waitForPublishedEvents` — why?
+  string
```

Comment:
Hmm, I'm curious what is going on here

Reply:
If I change this to `NetworkControllerEventType`, I get the following error in `waitForPublishedEvents` at the callsites for `subscribe` and `unsubscribe`:

`[tsserver 2345] [E] Argument of type '(...args: E["payload"]) => void' is not assignable to parameter of type 'ExtractEventHandler<NetworkControllerNetworkDidChangeEvent, E["type"]> | ExtractEventHandler<NetworkControllerNetworkWillChangeEvent, E["type"]> | ExtractEventHandler<...> | ExtractEventHandler<...>'.`

If I change 

``` typescript
let eventListener: (...args: E['payload']) => void;
```

to

``` typescript
let eventListener: ExtractEventHandler<E, E['type']>;
```

then I don't get that error at the `subscribe` or `unsubscribe` callsites, but I do get an error when attempting to set `eventListener`:

`[tsserver 2322] [E] Type '(...payload: E['payload']) => void' is not assignable to type 'ExtractEventHandler<E, E["type"]>'.`

And using `string` as I do now makes that error go away.

I feel like this is how RestrictedControllerMessenger was designed to be used. You can specify which events the messenger officially recognizes, but then you can subscribe and unsubscribe to any event: https://github.com/MetaMask/core/blob/a3b454eadbc9e974a6ee4a2729baf54171d422ed/packages/base-controller/src/ControllerMessenger.ts#L285. Perhaps this is to account for this library being used in a JavaScript context. In any case, because of this, it needs to infer the type of the event handler, and I guess that doesn't work very well? Based on the fact that TypeScript is expanding `ExtractEventHandler<E, E["type"]>` to `ExtractEventHandler<NetworkControllerNetworkDidChangeEvent, E["type"]> | ExtractEventHandler<NetworkControllerNetworkWillChangeEvent, E["type"]> | ExtractEventHandler<...> | ExtractEventHandler<...>`, it makes me think that something is disconnected in the types. I'm just not sure what.

I guess this isn't a big deal, I can remove the TODO, but it is slightly bothersome.

Reply:
The restricted controller messenger was designed to be used solely within the controller itself. That seems like the problem here - the tests shouldn't be using it at all. In tests we can use the top-level restricted messenger. It was definitely not designed to allow subscribing to any event.

Reply:
Sorry, something's not connecting for me. In order to use the NetworkController, we have to pass a RestrictedControllerMessenger that recognizes NetworkController events. So we can't make a generic restricted messenger, can we? 

I feel like we have a lot of bad examples in the core repo and elsewhere on how to do this properly. Is there a good example I can use?

Reply:
Right, so we need to create a restricted messenger to pass to the controller, but we create that from the top-level messenger. We can use that top-level messenger in tests.

Reply:
Here: https://github.com/MetaMask/metamask-extension/pull/18476/files#diff-89fa1edc4a981c9e4d156ed188d409ca35653a1de7ef939648f64ca7be9cbb03R7672
We're constructing a top-level messenger then throwing it away, returning just the restricted one. Instead we should be using the top-level messenger in our test helpers, and using the restricted one just for the constructor options

Reply:
I don't know how to explain that type error you encountered though. I'm unsure whether switching to the root messenger would solve that.

Reply:
Oh, right, because the restricted controller messenger modifies the top-level messenger, so the effects of the restricted controller messenger can be accessed via the top-level messenger. Does that sound correct?

Reply:
Well, it routes everything to the top-level messenger. The restricted messenger is a proxy for the top-level one that makes TypeScript complain if you try using something you aren't allowed to.

Reply:
Okay, I'm now using the unrestricted messenger here: [627c0a4](https://github.com/MetaMask/metamask-extension/pull/18476/commits/627c0a4a2c7e7c13e91480b4c14eb7e3fc556da5). This is a better practice, but unfortunately I'm still getting the error above. It seems like the unrestricted version of `subscribe` [tries to infer the payload type from the given event](https://github.com/MetaMask/core/blob/a3b454eadbc9e974a6ee4a2729baf54171d422ed/packages/base-controller/src/ControllerMessenger.ts#L488). However, I'm not sure why, as we know the given event type is one of the possible event types! It seems odd that the method signature is not this:

``` typescript
subscribe<E>(
  eventType: E['type'],
  handler: (...payload: E['payload') => void,
): void;
```

Anyway, for now I've just marked `eventListener` as `any`, and I'll investigate this further later.

Reply:
I think I've found a minimal example of this problem: https://www.typescriptlang.org/play?#code/C4TwDgpgBAogHgQwLZgDYRgNwgO2FAXigG8AoKKMBEVAewQBMAuKAZ2ACcBLHAcwBpSAXwDcpUqEiw4nBAGNgABWp1GAHiy5gAPkKxseKBBm4GrEuUor6zKDwBmEDlEVihlgPwvLLHBGwcYqT2AK44Cly0OFByHBAIwBAAEgg4DOgcGkYmaebwyGgYBjoAFFQ0NiwwANoA5OWqDLUAugCUVTIc8krW6jC6ZBRxwCEc0Q02bqRAA

It seems that TypeScript doesn't know how to reconcile a generic type with a conditional type.

Using `any` seems like the best plan for now, but we may be able to fix this by reducing the use of conditional types in the controller messenger.